api/src/main/java/org/apache/iceberg/types/PruneColumns.java

```diff
@@ -77,43 +88,48 @@ public Type struct(Types.StructType struct, List<Type> fieldResults) {
   @Override
   public Type field(Types.NestedField field, Type fieldResult) {
     if (selected.contains(field.fieldId())) {
-      return field.type();
+      if (field.type().isStructType() && !selectFullTypes) {
+        return projectSelectedStruct(fieldResult);
+      } else {
+        Preconditions.checkArgument(selectFullTypes || !field.type().isNestedType(),
+            "Cannot select partial List or Map types explicitly, %s:%s of type %s was selected",
+            field.fieldId(), field.name(), field.type());
+        // Selected non-struct field
+        return field.type();
+      }
     } else if (fieldResult != null) {
-      // this isn't necessarily the same as field.type() because a struct may not have all
-      // fields selected.
-      return fieldResult;
+      // This field wasn't selected but a subfield was so include that
+      return  fieldResult;
     }
     return null;
   }
 
   @Override
   public Type list(Types.ListType list, Type elementResult) {
     if (selected.contains(list.elementId())) {
-      return list;
-    } else if (elementResult != null) {
-      if (list.elementType() == elementResult) {
-        return list;
-      } else if (list.isElementOptional()) {
-        return Types.ListType.ofOptional(list.elementId(), elementResult);
+      if (list.elementType().isStructType() && !selectFullTypes) {
```

Comment:
I think this should be slightly different:

```java
if (selectFullTypes) {
  return list;
} else if (list.elementType().isStructType()) {
  StructType projectedStruct = projectSelectedStruct(elementResult);
  return projectList(list, projectedStruct);
} else {
  Precondtions.checkArgument(list.elementType().isPrimitive(), "Cannot directly project list or map");
  return list;
}
```

I think that makes it more understandable because `selectFullTypes` is handled first and so we know that the remaining cases are not selecting full types. It also handles lists of lists because of the argument check. I think that this was allowing an inner list to be selected directly.

Reply:
You should probably add test cases for list of lists or map of maps.